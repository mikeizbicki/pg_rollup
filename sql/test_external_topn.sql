create or replace language plpython3u;
create extension if not exists topn;
drop extension pg_rollup;
create extension if not exists pg_rollup;

create temporary table test_topn (
    id serial primary key,
    name text,
    num int
);

insert into test_topn (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);

select create_rollup(
    'test_topn',
    'test_topn_rollup1',
    wheres => 'num',
    rollups => $$
        topn(name)
    $$
);

select assert_rollup('test_topn_rollup1');

insert into test_topn (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);


select assert_rollup('test_topn_rollup1');
