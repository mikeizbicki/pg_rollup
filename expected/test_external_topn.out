SET client_min_messages TO WARNING;
create table test_topn (
    id serial primary key,
    name text,
    num int
);
insert into test_topn (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
CREATE MATERIALIZED VIEW test_topn_rollup1 AS (
    SELECT
        topn(topn_add_agg(name),1) AS top1,
        topn(topn_add_agg(name),2) AS top2,
        topn(topn_add_agg(name),3) AS top3
    FROM test_topn
);
select assert_rollup('test_topn_rollup1');
 assert_rollup 
---------------
 
(1 row)

insert into test_topn (name,num) values
    ('alice', 1),
    ('alice', 2),
    ('alice', 3),
    ('alice', 4),
    ('alice', 5),
    ('bill', 5),
    ('bill', 5),
    ('bill', 5),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 1),
    ('charlie', 3),
    ('charlie', NULL),
    ('dave', 4),
    ('elliot', 5),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 1),
    (NULL, 8),
    (NULL, 9),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL),
    (NULL, NULL);
select assert_rollup('test_topn_rollup1');
 assert_rollup 
---------------
 
(1 row)

select * from test_topn_rollup1;
    top1    |     top2     |     top3     
------------+--------------+--------------
 (alice,10) | (alice,10)   | (alice,10)
            | (charlie,10) | (charlie,10)
            |              | (bill,6)
(3 rows)

drop table test_topn cascade;
